libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.s
+
+import { DeclarationReference, Navigation } from '@microsoft/tsdoc/lib-commonjs/beta/DeclarationReference';
+import { ApiItem, IApiItemJson, IApiItemConstructor, IApiItemOptions } from '../items/ApiItem';
+import { DeserializerContext } from '../model/DeserializerContext';
+
+/**
+ * Constructor options for {@link (IApiExportedMixinOptions:interface)}.
+ * @public
+ */
+export interface IApiExportedMixinOptions extends IApiItemOptions {
+  isExported: boolean;
+}
+
+export interface IApiExportedMixinJson extends IApiItemJson {
+  isExported: boolean;
+}
+
+const _isExported: unique symbol = Symbol('ApiExportedMixin._isExported');
+
+/**
+ * The mixin base class for API items that can be exported.
```

Comment:
What qualifies as "exported"?

**src/index.ts** (entry point for API Extractor)
```ts
export { C } from './file';
```

**src/file.ts**
```ts
// C is "exported"
export class C { }

// D is exported by file.ts, but it is not exported by the package.
// Is D "exported"?
export class D { }

// n is not "exported"
namespace n {
  // Is f() "exported"?
  export function f(): void { }
}
```

Maybe we need a different jargon, for example "importable"?

Reply:
> What qualifies as "exported"?

An item is exported if it is exported from the package, that is, exported from the package's single entry point. Local exports (e.g. `D`, `n.f`) are not exported from the package and thus are not "exported". They might still be exported from a local file or a local non-exported namespace.

Another option is "consumable", but that doesn't quite work because if you have:

```
export class A {
  public prop: number;
}
```

I think most would interpret `A.prop` as consumable, whereas in our usage it's not the case.

Another pro for `ApiExportedMixin` over `ApiImportableMixin` is that it aligns with the `Navigation.Exports` navigation step used by TSDoc (i.e. an item has `isExported` set to true IFF its navigation step is `Navigation.Exports`).

@wcauchois do you have any opinions?

Reply:
I understand. However for a novice, it might be challenging to understand under what conditions `isExported=false` for a declaration that conspicuously includes the word `export`.  

On the other hand, an API consumer can import `C` using `import { C } from 'library'`, but they cannot use `import` with any of the other examples `A.prop`, `D`, `n`, `n.f`.  "Importable" might not be the best terminology, but maybe it shows that this concept could be simplified.

Reply:
So ultimately, the purpose of `ApiExportedMixin` and the `ApiExportedMixin.isExported` property is two-fold:

#### 1. Allows api-extractor-model to construct the correct reference for a given API item.

api-extractor-model needs to know whether to write a `Navigation.Exports` or `Navigation.Local` step when constructing an API item's reference. Otherwise, the reference constructed for an item in a reference token won't match the reference constructed for an item in the doc model, and thus reference resolution won't work. Today, the navigation step chosen when constructing the reference in a reference token is done by the `DeclarationReferenceGenerator._getNavigationToSymbol` method, while the navigation step chosen when constructing the reference for an item is hardcoded in api-extractor-model in the item's `buildCanonicalReference` method (typically to `Navigation.Exports` for the top-level items we care about).

This hardcoding of `Navigation.Exports` worked in the past because we only had exported items in the doc model (this actually isn't quite true, unexported items in namespaces can still be included in the doc model, but we can ignore this for now, this is a bug AFAICT). But now that we're including unexported items in the doc model under certain conditions, we need a way to set the appropriate navigation step.

Whatever mixin/logic we use to do so, it needs to as closely match the current logic for setting navigation steps in `DeclarationReferenceGenerator._getNavigationToSymbol`. Whatever definition of "exported" is being used there to set `Navigation.Exports` should be the same definition that is used in our new mixin/logic.

#### 2. Allows a documentation tool such as api-documenter to show on an API item's documentation page whether or not that item is "exported" (for some definition of the word).

It's useful for users to know whether an API item is "consumable/importable" or "exported from the current package". In https://github.com/microsoft/rushstack/issues/3513#issuecomment-1191791900, @wcauchois proposed including or omitting the `export` keyword in the API item's signature to communicate this. The problem here is that as @octogonz stated it's possible for an API item to be exported from a file but not actually exported from the package's single entry point. Should those items get `export` keywords in their signatures?

This proposal might also be too subtle for users (many users might miss this). I'm envisioning a more prominent flag or blurb somewhere on the API item's page that indicates whether someone can import/consume it.

Reply:
The problem with `ApiImportableMixin` is the following setup:

```
/** @public */
export namespace n {
  export class Foo {}
}
```

The declaration reference for `Foo` should be something like `package!n.Foo:class`. `n` uses the `Navigation.Exports` step and `Foo` uses the `Navigation.Exports` step. However, `Foo` uses the `Navigation.Exports` step because it's exported from `n`, not because it's importable (which it isn't, you can't import `Foo`).

So I think I misspoke above when I said:

> > What qualifies as "exported"?
>
> An item is exported if it is exported from the package, that is, exported from the package's single entry point.

Instead, I'm trying to use the same definition of "exported" as is used by the `DeclarationReferenceGenerator._getNavigationToSymbol` when determining what navigation step to use. The definition that I believe is being used there is: _"An item is exported if it's exported from the package or exported from a namespace within the package."_ So for a scenario like this:

```
namespace n {
  export class Foo {}
}
```

`n` is not exported, but `Foo` is exported. Whether or not an item is exported is determined with respect to its parent "module" (I think that's right?).

Note that this definition of "exported" is less useful to surface on a documentation site. A user doesn't care whether or not `Foo` is exported from `n`, they care whether `Foo` is "consumable". Instead, we'd probably want to expose a function in api-extractor-model that you can use to determine whether an item is consumable by checking if it and its entire parent tree is exported.

Reply:
> > What qualifies as "exported"?
>
> An item is exported if it is exported from the package, that is, exported from the package's single entry point. Local exports (e.g. D, n.f) are not exported from the package and thus are not "exported". They might still be exported from a local file or a local non-exported namespace.

I now realize that the definition I provided here and the one I'm using in this PR are not consistent with what the `Navigation.Exports` step represents. The two need to be consistent, because I've implemented that an item has a `Navigation.Exports` step if it is exported.

In `DeclarationReferenceGenerator._getNavigationToSymbol`, an item seems to have a `Navigation.Exports` step if it is exported from its parent module (i.e. either the package entry point or a local namespace). Whereas in this PR, I've implemented that an item has a `Navigation.Exports` step if `ApiExportedMixin.isExported` is true, and `ApiExportedMixin.isExported` is true if the item is exported from the package. So I'll need to iterate on this a bit... 🤔 

Perhaps we should return to the previous words _consumable_ and _exported_, albeit with different definitions from before:

### Before this PR

* *Exported:* A top-level export.
* *Consumable:* Either a top-level export or exported from an exported namespace.

### After

* *Exported:* An item exported from its parent module (i.e. either package entry point or local namespace).
* *Consumable:* Same as above.

So in the examples below:

```
export class A {}
class B {}

export namespace n1 {
  export class C {}
  class D {}
}

namespace n2 {
  export class E {}
  class F {}
}
```

### Before this PR

| Item | Exported | Consumable |
| --- | --- | --- |
| A | Yes | Yes |
| B | No | No |
| C | No | Yes |
| D | No | No |
| E | No | No |
| F | No | No |

### After

| Item | Exported | Consumable |
| --- | --- | --- |
| A | Yes | Yes |
| B | No | No |
| C | Yes | Yes |
| D | No | No |
| E | Yes | No |
| F | No | No |

With these new definitions, we can write the appropriate navigation step.

Reply:
I updated the PR to use these new definitions, thus addressing the inconsistency between what `ApiExportedMixin.isExported` means and what `Navigation.Exports` means. Also updated the original PR description.